server/src/main/java/org/apache/iotdb/db/metadata/template/Template.java

```diff
@@ -177,6 +208,406 @@ public String getMeasurementNodeName(String measurementName) {
+      CompressionType[] compressors)
+      throws IllegalPathException {
+    String prefix;
+    String[] pathNode;
+
+    // deduplicate
+    Set<String> pathSet = new HashSet<>(Arrays.asList(measurements));
+    if (pathSet.size() != measurements.length)
+      throw new IllegalPathException("Duplication in paths.");
+
+    for (int i = 0; i <= measurements.length - 1; i++) {
+      pathNode = MetaUtils.splitPathToDetachedPath(measurements[i]);
+
+      // If prefix exists and aligned, it will throw exception
+      prefix = joinBySeparator(Arrays.copyOf(pathNode, pathNode.length - 1));
+      if ((isPathExist(prefix) != null) && (alignedPrefix.contains(prefix)))
+        throw new IllegalPathException(prefix, "path already exists and aligned");
+
+      IMeasurementSchema schema =
+          constructSchema(
+              pathNode[pathNode.length - 1], dataTypes[i], encodings[i], compressors[i]);
+      constructTemplateTree(measurements[i], schema);
+    }
+  }
+
+  // endregion
+
+  // region deduction of template
+
+  public void deleteMeasurements(String path) throws IllegalPathException {
+    IMNode cur = isPathExist(path);
+    if (cur == null) throw new IllegalPathException(path, "Path does not exist");
+    if (!cur.isMeasurement())
+      throw new IllegalPathException(path, "Path is not pointed to a measurement node.");
+
+    IMNode par = cur.getParent();
+    par.deleteChild(cur.getName());
+  }
+
+  public void deleteSeriesCascade(String path) throws IllegalPathException {
+    IMNode cur = isPathExist(path);
+    IMNode par;
+
+    if (cur == null) throw new IllegalPathException(path, "Path not exists.");
+    par = cur.getParent();
+    par.deleteChild(cur.getName());
+
+    // Remove all aligned prefix below the series path
+    Deque<IMNode> astack = new ArrayDeque<>();
+    astack.push(cur);
+    while (astack.size() != 0) {
+      IMNode top = astack.pop();
+      if (!top.isMeasurement()) {
+        String thisPrefix = getFullPathWithoutTemplateName(top);
+        if (alignedPrefix.contains(thisPrefix)) alignedPrefix.remove(thisPrefix);
+        for (IMNode child : top.getChildren().values()) {
+          astack.push(child);
+        }
+      }
+    }
+  }
+
+  public void deleteAlignedPrefix(String path) {
+    if (alignedPrefix.contains(path)) alignedPrefix.remove(path);
+  }
+  // endregion
+
+  // region unimplemented interfaces
+  public boolean updateSeriesDataType(String path, TSDataType dataType) {
+    return true;
+  }
+
+  public boolean updateSeriesEncoding(String path, TSEncoding encoding) {
+    return true;
+  }
+
+  public boolean updateSeriesCompression(String path, CompressionType compressor) {
+    return true;
+  }
+
+  public boolean updateAlignedPrefixCompression(String prefix, CompressionType compressor) {
+    return true;
+  }
+
+  public boolean upsertSeries(
+      String path, TSDataType dataType, TSEncoding encoding, CompressionType compressionType) {
+    return true;
+  }
+  // endregion
```

Comment:
I'don't think we need to support schema modification. If users want to alter the schema of one timeseries, they should delete target timeseries first and create what they want.

Reply:
all removed